subsys/debug/ppi_trace/ppi_trace.c

```diff
@@ -0,0 +1,220 @@
+	} else {
+		err = nrfx_dppi_channel_alloc(ch);
+	}
+#else
+	err = nrfx_ppi_channel_alloc((nrf_ppi_channel_t *)ch);
+#endif
+	return err;
+}
+
+/** @brief Set task and event on (D)PPI channel. */
+static void ppi_assign(u8_t ch, u32_t evt, u32_t task)
+{
+#ifdef DPPI_PRESENT
+	/* Use mask of one of subscribe registers in the system, assuming that
+	 * all subscribe registers has the same mask for enabling channel.
+	 */
+	*SUBSCRIBE_ADDR(task) = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | (u32_t)ch;
+	*PUBLISH_ADDR(evt) = DPPIC_SUBSCRIBE_CHG_EN_EN_Msk | (u32_t)ch;
+#else
+	(void)nrfx_ppi_channel_assign(ch, evt, task);
+#endif
+}
+
+/** @brief Enable (D)PPI channels. */
+static void ppi_enable(u32_t channel_mask)
+{
+#ifdef DPPI_PRESENT
+	nrf_dppi_channels_enable(channel_mask);
+#else
+	nrf_ppi_channels_enable(channel_mask);
+#endif
+}
+
+/** @brief Disable (D)PPI channels. */
+static void ppi_disable(u32_t channel_mask)
+{
+#ifdef DPPI_PRESENT
+	nrf_dppi_channels_disable(channel_mask);
+#else
+	nrf_ppi_channels_disable(channel_mask);
+#endif
+}
+
+/** @brief Allocate GPIOTE channel.
+ *
+ * @param pin Pin.
+ *
+ * @return Allocated channel or -1 if failed to allocate.
+ */
+static int gpiote_channel_alloc(u32_t pin)
+{
+	for (u8_t channel = 0; channel < GPIOTE_CH_NUM; ++channel) {
+		if (!nrf_gpiote_te_is_enabled(channel)) {
+			nrf_gpiote_task_configure(channel, pin,
+						  NRF_GPIOTE_POLARITY_TOGGLE,
+						  NRF_GPIOTE_INITIAL_VALUE_LOW);
+			nrf_gpiote_task_enable(channel);
+			return channel;
+		}
+	}
+
+	return -1;
+}
+
+void *ppi_trace_config(u32_t pin, u32_t evt)
+{
+	int err;
+	u32_t task;
+	int gpiote_ch;
+	nrf_gpiote_tasks_t task_id;
+	u8_t ppi_ch;
+
+	err = ppi_alloc(&ppi_ch, evt);
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("Failed to allocate PPI channel.");
+		return NULL;
+	}
+
+	gpiote_ch = gpiote_channel_alloc(pin);
+	if (gpiote_ch < 0) {
+		LOG_ERR("Failed to allocate GPIOTE channel.");
+		return NULL;
+	}
+
+	task_id = offsetof(NRF_GPIOTE_Type, TASKS_OUT[gpiote_ch]);
+	task = nrf_gpiote_task_addr_get(task_id);
+	ppi_assign(ppi_ch, evt, task);
+
+	return HANDLE_ENCODE(ppi_ch);
+}
+
+void *ppi_trace_pair_config(u32_t pin, u32_t start_evt, u32_t stop_evt)
+{
+	int err;
+	u32_t task;
+	u8_t start_ch;
+	u8_t stop_ch;
+	int gpiote_ch;
+	nrf_gpiote_tasks_t task_set_id;
+	nrf_gpiote_tasks_t task_clr_id;
+
+	if (!GPIOTE_SET_CLR_PRESENT) {
+		__ASSERT(0, "Function not supported on this platform.");
+		return NULL;
+	}
+
+
+	err = ppi_alloc(&start_ch, start_evt);
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("Failed to allocate PPI channel.");
+		return NULL;
+	}
+
+	err = ppi_alloc(&stop_ch, stop_evt);
+	if (err != NRFX_SUCCESS) {
+		LOG_ERR("Failed to allocate PPI channel.");
+		return NULL;
+	}
+
+	gpiote_ch = gpiote_channel_alloc(pin);
+	if (gpiote_ch < 0) {
+		LOG_ERR("Failed to allocate GPIOTE channel.");
+		return NULL;
+	}
+
+	task_set_id = COND_CODE_1(GPIOTE_SET_CLR_PRESENT,
+			(offsetof(NRF_GPIOTE_Type, TASKS_SET[gpiote_ch])), (0));
```

Comment:
I still don't get it why complicate the code that much. When `GPIOTE_SET_CLR_PRESENT` is not active, this part will not be executed at all. Actually, the compiler should be able to get rid of all the code after the first `return NULL`. Why define `GPIOTE_SET_CLR_PRESENT` then? Wouldn't it be easier to use something like I [suggested](https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/623#discussion_r288428366) earlier? Or do I still miss something that excludes such solution?

Reply:
done